src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,168 +23,324 @@ using namespace opossum::expression_functional;  // NOLINT
+  // uniqueness is not guaranteed, we cannot perform the rewrite as non-unique joins could possibly emit a matching
+  // line more than once.
+
+  auto join_node = std::dynamic_pointer_cast<JoinNode>(node);
+  if (join_node->join_mode != JoinMode::Inner) return;
+
+  // Check whether the left/right inputs are actually needed by following operators
+  auto left_input_is_used = false;
+  auto right_input_is_used = false;
+  for (const auto& output : node->outputs()) {
+    for (const auto& required_expression : required_expressions_by_node[output]) {
+      if (expression_evaluable_on_lqp(required_expression, *node->left_input())) left_input_is_used = true;
+      if (expression_evaluable_on_lqp(required_expression, *node->right_input())) right_input_is_used = true;
     }
+  }
+  DebugAssert(left_input_is_used || right_input_is_used, "Did not expect a useless join");
+  if (left_input_is_used && right_input_is_used) return;
+
+  // Check whether the join predicates operate on unique columns.
+  auto left_joins_on_unique_column = false;
+  auto right_joins_on_unique_column = false;
+
+  const auto& join_predicates = join_node->join_predicates();
+  for (const auto& join_predicate : join_predicates) {
+    const auto is_unique_column = [](const auto& expression) {
+      const auto& column = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+      if (!column) return false;
 
-    if (const auto stored_table_node = std::dynamic_pointer_cast<StoredTableNode>(node)) {
-      stored_table_node->set_pruned_column_ids(pruned_column_ids);
-    } else if (const auto mock_node = std::dynamic_pointer_cast<MockNode>(node)) {
-      mock_node->set_pruned_column_ids(pruned_column_ids);
-    } else {
-      // We don't know how to prune columns from this leaf-node type (CreateViewNode, etc.), so do nothing
+      const auto& column_reference = column->column_reference;
+      const auto& stored_table_node =
+          std::dynamic_pointer_cast<const StoredTableNode>(column_reference.original_node());
+      if (!stored_table_node) return false;
+
+      const auto& table = Hyrise::get().storage_manager.get_table(stored_table_node->table_name);
+      for (const auto& table_constraint : table->get_soft_unique_constraints()) {
+        // This currently does not handle multi-column constraints, but that should be easy to add once needed.
+        if (table_constraint.columns.size() > 1) continue;
+        if (table_constraint.columns[0] == column_reference.original_column_id()) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    const auto& predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(join_predicate);
+    const auto& left_operand = predicate->left_operand();
+    const auto& right_operand = predicate->right_operand();
+
+    if (!left_joins_on_unique_column) {
+      left_joins_on_unique_column = is_unique_column(left_operand);
```

Comment:
Can't follow this loop. You iterate over the join predicates and overwrite the `*joins_on_unique_column` bool every time. What if the primary join is an optimizable join, but the secondary predicate is not? That's probably the good case where nothing breaks.
I am wondering if there are cases where secondary predicates might introduce serious problems.

Reply:
> You iterate over the join predicates and overwrite the *joins_on_unique_column bool every time.

No, see the `left_joins_on_unique_column` above. But `|=` is clearer.

> What if the primary join is an optimizable join, but the secondary predicate is not? That's probably the good case where nothing breaks.

Added